ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,1319 @@
+
+Notes:
+* if the service task has an input variable mapping then it produces one or more `Variable:Created` events for the new local variables
+* if the job is completed with variables then these variables will be stored in the state until they are used for the output mapping (i.e. temporal/transient variables). They must therefore also be written into the `Job:Completed` event.
+* if the service task has an output variable mapping, or the job is completed with variables, then it produces one or more `Variable:Created`/`Variable:Updated` events for the new and updated variables of the workflow instance or scope
+* the order of writing follow-up events matters. Both `ServiceTask:Element_Activated` and `Variable:Created` need to be replayed before `Job:Created`.
+
+On processing:
+* the `ServiceTask:Activate_Element` command transitions the service task
+  element to activated, may create variables for input mapping and creates a job
+* the `JobBatch:Activate` command activates a batch of jobs
+* the `Job:Complete` command completes a job
+* the `ServiceTask:Complete_Element` command transitions the service task
+  element to completed, may create and/or update variables for output mapping
+  and marks the sequence flow as taken
+
+On replay:
+* the `ServiceTask:Element_Activating` event transitions the service task element to activating
+* the `Variable:Created` events create a variables for input mapping
+* the `ServiceTask:Element_Activated` event transitions the service task element to activated
+* the `Job:Created` event creates a job
+* the `JobBatch:Activated` event activates a batch of jobs
+* the `Job:Completed` event completes a job
+* the `Variable:Created` events create variables for output mapping
+* the `Variable:Updated` events update variables for output mapping
+* the `ServiceTask:Element_Completing` event transitions the service task element to completing
+* the `ServiceTask:Element_Completed` event transitions the service task element to completed
+* the `Sequence_Flow_Taken` event marks the sequence flow as taken
+* processing can continue again at `EndEvent:Activate_Element`
+
+Differences to current processing:
+* the event `Job:Created` is produced by the command `ServiceTask:Activate_Element` instead of a `Job:Create` command because it can't be interrupted
+
+Considered alternatives:
+* just like call activity variable propagation we could consider storing job
+  variables in `TemporaryVariable:Created` events
+
+### Message Intermediate Catch Event
+
+![message-correlation](ZEP-wf-stream-processing/event-sourcing-message-correlation-1.png)
+
+![message-correlation](ZEP-wf-stream-processing/event-sourcing-message-correlation-2.png)
+
+Command to Process | Written Events and Follow-up Commands
+---|---
+IntermediateCatchEvent:Activate_Element | IntermediateCatchEvent:Element_Activating, IntermediateCatchEvent:Element_Activated, MessageSubscription:Open
```

Comment:
@saig0 Could we add `WorkflowInstanceSubscription:Opened` event here and remove the Open command from the `MessageSubscription:Open` processing? That would be more in line with Timers/Jobs/etc.

Reply:
This is not possible due to the dependency on `MessageSubScription:Opened` on another partition. I'll expand the description to explain why